File: src/main/java/dev/psygamer/wireframe/api/item/ItemAttributes.java
package dev.psygamer.wireframe.api.item;

import dev.psygamer.wireframe.core.impl.Instancer;

import net.minecraft.item.Item;
import net.minecraft.item.ItemGroup;
import net.minecraft.item.Rarity;

public abstract class ItemAttributes {
	
	private static final ItemAttributes INSTANCE = Instancer.createInstance();
	
	/**
	 * @return A new instance of a {@link ItemAttributes}
	 * @author psyGamer
	 * @version 1.0 | Minecraft 1.16 +
	 * @since 1.0 | Minecraft 1.16
	 */
	public static ItemAttributes create() {
		return INSTANCE.createInstance();
	}
	
	/**
	 * @param group Creative Tab for the Block
	 * @return A new instance of a {@link ItemAttributes}
	 * @author psyGamer
	 * @version 1.0 | Minecraft 1.16 +
	 * @since 1.0 | Minecraft 1.16
	 */
	public static ItemAttributes create(final ItemGroup group) {
		return INSTANCE.createInstance(group);
	}
	
	public abstract ItemAttributes maxStackSize(final int stackSize);
	
	public abstract ItemAttributes rarity(final Rarity rarity);
	
	public abstract ItemAttributes fireResistant(final boolean fireResistant);
	
	public abstract ItemAttributes repairable(final boolean repairable);
	
	public abstract ItemAttributes craftingRemainder(final Item remainder);
	
	protected abstract ItemAttributes createInstance();
	
	protected abstract ItemAttributes createInstance(final ItemGroup group);
}

File: src/main/kotlin/dev/psygamer/wireframe/api/item/ItemAttributes.kt
package dev.psygamer.wireframe.api.item

import dev.psygamer.wireframe.nativeapi.item.NativeItemAttributes
import dev.psygamer.wireframe.api.item.util.Rarity
import net.minecraft.item.Item
import net.minecraft.item.ItemGroup

/**
 * A factory class for easily creating attributes related to items.
 *
 * @param itemGroup Creative Tab for the Block
 * @see Item
 */
class ItemAttributes @JvmOverloads constructor(var itemGroup: ItemGroup? = null) {
	
	var maxStackSize = 64
		private set
	var isFireResistant = false
		private set
	var isRepairable = false
		private set
	var craftingRemainder: Item? = null
		private set
	var rarity = Rarity.COMMON
		private set
	
	internal val mcNative: NativeItemAttributes = NativeItemAttributes(this)
	
	fun maxStackSize(maxStackSize: Int): ItemAttributes {
		this.maxStackSize = maxStackSize
		return this
	}
	
	fun fireResistant(fireResistant: Boolean): ItemAttributes {
		isFireResistant = fireResistant
		return this
	}
	
	fun repairable(repairable: Boolean): ItemAttributes {
		isRepairable = repairable
		return this
	}
	
	fun craftingRemainder(craftingRemainder: Item?): ItemAttributes {
		this.craftingRemainder = craftingRemainder
		return this
	}
	
	fun rarity(rarity: Rarity): ItemAttributes {
		this.rarity = rarity
		return this
	}
}